[popover] Prevent hidePopover from causing a recursive loop

rwlbuis opened the pull request:
#### 0bd9e48790d43285feb2f971e6b5ae12dc290fc8
<pre>
[popover] Prevent hidePopover from causing a recursive loop
<a href="https://bugs.webkit.org/show_bug.cgi?id=256379">https://bugs.webkit.org/show_bug.cgi?id=256379</a>

Reviewed by Tim Nguyen.

Fix hideAllPopoversUntil for the case where hiding an auto popover changes the auto popover
list (through beforetoggle) so the same auto popover is on top, causing a loop. To fix this, when
this is detected, continue the loop without firing events.

See:
<a href="https://github.com/whatwg/html/pull/9198">https://github.com/whatwg/html/pull/9198</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::topmostAutoPopover const):
(WebCore::Document::hideAllPopoversUntil):
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::ScopedStartShowingOrHiding::ScopedStartShowingOrHiding):
(WebCore::PopoverData::ScopedStartShowingOrHiding::~ScopedStartShowingOrHiding):
(WebCore::PopoverData::ScopedStartShowingOrHiding::wasShowingOrHiding const):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::hidePopoverInternal):

Canonical link: <a href="https://commits.webkit.org/264623@main">https://commits.webkit.org/264623@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe61382e6d6d7063b3ab2b2ce3e09d6a062e905f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8089 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8590 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9744 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8185 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8094 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10362 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8287 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11044 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8233 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9315 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7369 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9865 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6616 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7407 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14975 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7739 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7533 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10882 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7999 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6509 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7309 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7320 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1965 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11518 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7741 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9014,31 +9014,41 @@ HTMLDialogElement* Document::activeModalDialog() const
-        if (element.popoverState() == PopoverState::Auto) {
-            if (&element == endpoint)
+    auto closeAllOpenPopovers = [&]() {
+        while (auto* popover = topmostAutoPopover())
```
Would be nice to use `RefPtr` for `popover`.

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9014,31 +9014,41 @@ HTMLDialogElement* Document::activeModalDialog() const
+        bool foundEndPoint = false;
+        RefPtr<HTMLElement> lastToHide;
+        for (auto& popover : stack) {
+            if (popover.get() == endpoint) {
```
We don't use curly brackets for one-liners.

rwlbuis replied:
Removed.

cdumez commented on Source/WebCore/dom/PopoverData.h:
```diff
@@ -62,12 +62,34 @@ class PopoverData {
     void setInvoker(const HTMLFormControlElement* element) { m_invoker = element; }
 
+    class ScopedStartShowingOrHiding {
+        WTF_MAKE_FAST_ALLOCATED;
```
Seems like a class we wouldn't want to heap-allocate so I would omit `WTF_MAKE_FAST_ALLOCATED`.

rwlbuis replied:
Removed.

cdumez commented on Source/WebCore/dom/PopoverData.h:
```diff
@@ -62,12 +62,34 @@ class PopoverData {
+    explicit operator bool() const { return m_wasSet; }
+
+    private:
+        const Element& m_popover;
```
Would be safer to Ref the popover, since this is meant to be used as a stack variable, there is no risk of reference cycle.

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1391,9 +1396,11 @@ ExceptionOr<void> HTMLElement::showPopover(const HTMLFormControlElement* invoker
             return { };
+
+        shouldRestoreFocus = !document->topmostAutoPopover();
+        document->popoverStack().append(this);
```
`*this`

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/dom/Document.h:
```diff
@@ -2219,6 +2221,8 @@ class Document
 
     ListHashSet<Ref<Element>> m_topLayerElements;
 
+    Vector<RefPtr<HTMLElement>> m_popoverStack;
```
I don't think the pointers can be null so we probably want to use `Ref<>`, not `RefPtr<>`.

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1449,6 +1461,9 @@ ExceptionOr<void> HTMLElement::hidePopoverInternal(FocusPreviousElement focusPre
     removeFromTopLayer();
 
+    if (popoverState() == PopoverState::Auto)
+        document().popoverStack().removeLast();
```
There are some early returned between you adding to the stack and removing from the stack. I am worried we may end up leaving items in the stack and leak documents since the stack holds a strong ref to Nodes.

rwlbuis replied:
I think the behaviour is the same as for m_topLayerElements in Document (add both in showPopover, remove both inn hidePopover) and we do not seem to hit the release assert for m_topLayerElements in Document::removedLastRef.

cdumez replied:
If you leak an HTMLElement then you leak its Document too and thus Document::removedLastRef() will never get called. As a result, I don't know what an assertion in Document::removedLastRef would prove.

I don't think we should land this patch as-is, too likely to introduce leaks.

cdumez replied:
You need to guarantee that any element added to the stack will be removed. Using some sort of scope variable would likely work.

nt1m replied:
@cdumez @rwlbuis I wonder if it's actually worth caching `popoverStack()` (which should probably be called `autoPopoverList()` fwiw) or not, maybe that method should just build a vector on the go of the filtered top layer, after all, the top layer should be a small vector in the vast majority of cases.

rwlbuis replied:
Top level and auto popover list are somewhat related:
https://html.spec.whatwg.org/#auto-popover-list

Anyway, I found removedFromAncestor takes guaranteed care of removing top layer for the Element but does not guarantee cleanup for the auto popover list, so I added some code to take care of that, hopefully that fixes the leak problem?

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1449,6 +1464,9 @@ ExceptionOr<void> HTMLElement::hidePopoverInternal(FocusPreviousElement focusPre
     removeFromTopLayer();
 
+    if (popoverState() == PopoverState::Auto)
+        document().autoPopoverList().removeLast();
```
If the element being removed from the top layer isn't the last one, this is wrong.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -447,8 +447,11 @@ Node::InsertedIntoAncestorResult HTMLElement::insertedIntoAncestor(InsertionType
+    if (popoverData()) {
         hidePopoverInternal(FocusPreviousElement::No, FireEvents::No);
+        if (popoverState() == PopoverState::Auto && !document().autoPopoverList().isEmpty() && document().autoPopoverList().last().ptr() == this)
+            document().autoPopoverList().removeLast();
```
What if it's not the last element?

nt1m reviewed (commented):
Honestly I don't see much point in caching the autoPopoverList and trying to maintain it, can we just compute it once at the call site? Shouldn't be costly.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9019,20 +9025,51 @@ HTMLElement* Document::topmostAutoPopover() const
+                continue;
+#endif
+            if (auto* candidate = dynamicDowncast<HTMLElement>(item.get()); candidate && candidate->popoverState() == PopoverState::Auto) {
+                if (!is<HTMLDialogElement>(candidate) || !downcast<HTMLDialogElement>(candidate)->isOpen())
```
I don't think we need all of these checks, quoting the spec:

> The [Document ](https://whatpr.org/html/9198/3d8fca2...cb6b5f6/dom.html#document)has an auto popover list , which is a [list ](https://infra.spec.whatwg.org/#list)of all the elements in the [Document ](https://whatpr.org/html/9198/3d8fca2...cb6b5f6/dom.html#document)'s [top layer ](https://fullscreen.spec.whatwg.org/#top-layer)whose [popover ](https://whatpr.org/html/9198/3d8fca2...cb6b5f6/popover.html#attr-popover)attribute is in the [auto ](https://whatpr.org/html/9198/3d8fca2...cb6b5f6/popover.html#attr-popover-auto-state)state, in the same order.

rwlbuis replied:
This is what chromium and my previous patch do implicitly by the way the autopopover vector is maintained. Not doing these checks cause failures and even crashes.

rwlbuis replied:
Ah, I guess you meant the checks in the lambda, not in general. Yes looks like we do not need them in the lambda, removed.

cdumez commented on Source/WebCore/dom/PopoverData.h:
```diff
@@ -62,12 +62,33 @@ class PopoverData {
     std::optional<PopoverToggleEventData> m_queuedToggleEventData;
     WeakPtr<Element, WeakPtrImplWithEventTargetData> m_previouslyFocusedElement;
     WeakPtr<HTMLFormControlElement, WeakPtrImplWithEventTargetData> m_invoker;
+    bool m_isHidingOrShowingPopover;
```
This is missing an initializer.

rwlbuis replied:
Added.

cdumez commented on Source/WebCore/dom/PopoverData.h:
```diff
@@ -62,12 +62,33 @@ class PopoverData {
+        if (!m_wasSet && m_popover->popoverData())
+            m_popover->popoverData()->m_isHidingOrShowingPopover = false;
+    }
+    explicit operator bool() const { return m_wasSet; }
```
I think this operator's meaning is a bit confusing. Maybe we could have an explicit getter like `wasShowingOrHiding()`.

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1362,6 +1362,9 @@ ExceptionOr<void> HTMLElement::showPopover(const HTMLFormControlElement* invoker
 
     ASSERT(!isInTopLayer());
 
+    PopoverData::ScopedStartShowingOrHiding scopedWasShowingOrHiding(*this);
```
Maybe we could rename this to `showOrHidingPopoverScope`.

rwlbuis replied:
Done.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1362,6 +1362,9 @@ ExceptionOr<void> HTMLElement::showPopover(const HTMLFormControlElement* invoker
     ASSERT(!isInTopLayer());
 
+    PopoverData::ScopedStartShowingOrHiding scopedWasShowingOrHiding(*this);
+    FireEvents fireEvents = scopedWasShowingOrHiding ? FireEvents::No : FireEvents::Yes;
```
```suggestion
    auto fireEvents = scopedWasShowingOrHiding ? FireEvents::No : FireEvents::Yes;
```

rwlbuis replied:
Done.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9019,20 +9025,51 @@ HTMLElement* Document::topmostAutoPopover() const
+    bool repeatingHide = false;
+    do {
+        bool foundEndPoint = false;
+        RefPtr<Element> lastToHide;
```
nit: Can we swap the two lines to match the spec?

rwlbuis replied:
Done.

rwlbuis committed:
[popover] Prevent hidePopover from causing a recursive loop
https://bugs.webkit.org/show_bug.cgi?id=256379

Reviewed by Tim Nguyen.

Fix hideAllPopoversUntil for the case where hiding an auto popover changes the auto popover
list (through beforetoggle) so the same auto popover is on top, causing a loop. To fix this, when
this is detected, continue the loop without firing events.

See:
https://github.com/whatwg/html/pull/9198

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::topmostAutoPopover const):
(WebCore::Document::hideAllPopoversUntil):
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::ScopedStartShowingOrHiding::ScopedStartShowingOrHiding):
(WebCore::PopoverData::ScopedStartShowingOrHiding::~ScopedStartShowingOrHiding):
(WebCore::PopoverData::ScopedStartShowingOrHiding::wasShowingOrHiding const):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::hidePopoverInternal):

Canonical link: https://commits.webkit.org/264623@main
